lib/logproto/logproto-proxied-text-server.c

```diff
@@ -0,0 +1,156 @@
+    int compares = -1;
+
+    matches = sscanf(input->str, PROXY_TCP_HDR, &self->info->ip_version, self->info->src_ip, self->info->dst_ip,
+                     &self->info->src_port, &self->info->dst_port);
+
+    if (matches != 5)
+        compares = strcmp(input->str, PROXY_UNKNOWN_HDR);
+
+    g_string_free(input, TRUE);
+
+    if (compares == 0) {
+        msg_info("PROXY UNKOWN valid protocol header received");
+        self->info->unknown = TRUE;
+    }
+
+    return (matches == 5 || compares == 0) ? TRUE : FALSE;
+}
+
+static void
+_log_proto_proxied_text_server_add_aux_data(LogProtoProxiedTextServer *self, LogTransportAuxData *aux) {
+    gchar buf1[8];
+    gchar buf2[8];
+    gchar buf3[8];
+
+    if (self->info->unknown)
+        return;
+
+    snprintf(buf1, 8, "%i", self->info->src_port);
+    snprintf(buf2, 8, "%i", self->info->dst_port);
+    snprintf(buf3, 8, "%i", self->info->ip_version);
+
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_SRCIP", self->info->src_ip);
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_DSTIP", self->info->dst_ip);
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_SRCPORT", buf1);
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_DSTPORT", buf2);
+    log_transport_aux_data_add_nv_pair(aux, "PROXIED_IP_VERSION", buf3);
+
+    return;
+}
+
+static LogProtoStatus
+_log_proto_proxied_text_server_handshake(LogProtoServer *s)
+{
+    const guchar *msg;
+    gsize msg_len;
+    gboolean may_read = TRUE;
+    gboolean parsable;
+    LogProtoStatus status;
+
+    LogProtoProxiedTextServer *self = (LogProtoProxiedTextServer *) s;
+
+    // Fetch a line from the transport layer
+    status = self->fetch_from_buffer(&self->super.super.super, &msg, &msg_len, &may_read, NULL, NULL);
+
+    if (status != LPS_SUCCESS)
+        return status;
+
+    parsable = _log_proto_proxied_text_server_parse_header(self, msg, msg_len);
+
+    msg_debug("PROXY protocol header received", evt_tag_printf("line", "%.*s", (gint) msg_len, msg));
+
+    if (parsable) {
+        msg_info("PROXY protocol header parsed successfully");
+        return LPS_SUCCESS;
+    } else {
+        msg_error("Error parsing PROXY protocol header");
+        return LPS_ERROR;
+    }
+}
+
+static gboolean
+_log_proto_proxied_text_server_handshake_in_progress(LogProtoServer *s)
+{
+    LogProtoProxiedTextServer *self = (LogProtoProxiedTextServer *) s;
+
+    // Run handshake() only once
+    if (!self->handshake_done)
+    {
+        self->handshake_done = TRUE;
+        return TRUE;
+    }
+    return FALSE;
+}
+
+static LogProtoStatus
+_log_proto_proxied_text_server_fetch(LogProtoServer *s, const guchar **msg, gsize *msg_len, gboolean *may_read,
+                                     LogTransportAuxData *aux, Bookmark *bookmark)
+{
+    LogProtoProxiedTextServer *self = (LogProtoProxiedTextServer *) s;
+
+    LogProtoStatus status = self->fetch_from_buffer(&self->super.super.super, msg, msg_len, may_read, aux, bookmark);
```

Comment:
I don't see how this works.
This method replaces `log_proto_buffered_server_fetch`, but does not call that method, so the input buffer will never be filled:

```
self->super.super.super.fetch = _log_proto_proxied_text_server_fetch;
```

----

Update: Oh I see, `fetch_from_buffer` is actually a pointer to `log_proto_buffered_server_fetch`, not to `log_proto_text_server_fetch_from_buffer` as I thought.

Can we publish `log_proto_buffered_server_fetch()` and call it directly, so there will be no confusion (or just rename this cached pointer to `fetch`)?

Reply:
done (it is more readable, thanks)